File: facecat_cpp/source/service/FCServerService.cpp
#include "..\\..\\stdafx.h"
#include "..\\..\\include\\service\\FCServerService.h"
#include "..\\..\\include\\core\\FCBinary.h"
#include "..\\..\\include\\sock\\FCServerSockets.h"

namespace FaceCat{
	long long m_downFlow = 0;

	long long m_upFlow = 0;

	MessageCallBack m_messageCallBack;

	map<int, FCServerService*> m_services;

	WriteLogCallBack m_writeLogCallBack;

	FCMessage::FCMessage(){
		m_groupID = 0;
		m_serviceID = 0;
		m_functionID = 0;
		m_sessionID = 0;
		m_requestID = 0;
		m_socketID = 0;
		m_state = 0;
		m_compressType = 0;
		m_bodyLength = 0;
		m_body = 0;
	}

	FCMessage::FCMessage(int groupID, int serviceID, int functionID, int sessionID, int requestID, int socketID, int state, int compressType, int bodyLength, char *body){
		m_groupID = groupID;
		m_serviceID = serviceID;
		m_functionID = functionID;
		m_sessionID = sessionID;
		m_requestID = requestID;
		m_socketID = socketID;
		m_state = state;
		m_compressType = compressType;
		m_bodyLength = bodyLength;
		m_body = body;
	}

	///////////////////////////////////////////////////////////////////////////////////////////////////////

	FCMessageListener::FCMessageListener(){
	}

	FCMessageListener::~FCMessageListener(){
		clear();
	}

	void FCMessageListener::add(ListenerMessageCallBack callBack, Object pInvoke){
		m_callBacks.push_back(callBack);
		m_callBackInvokes.push_back(pInvoke);
	}

	void FCMessageListener::callBack(FCMessage *message){
		int callBacksSize = (int)m_callBacks.size();
		for(int i = 0; i < callBacksSize; i++){
			m_callBacks[i](message, m_callBackInvokes[i]);
		}
	}

	void FCMessageListener::clear(){
		m_callBacks.clear();
		m_callBackInvokes.clear();
	}

	void FCMessageListener::remove(ListenerMessageCallBack callBack){
		Object pInvoke = 0;
		int pos = 0;
		vector<ListenerMessageCallBack>::iterator sIter = m_callBacks.begin();
		for(; sIter != m_callBacks.end(); ++sIter){
			if(callBack == *sIter)
			{
				m_callBacks.erase(sIter);
				pInvoke = m_callBackInvokes[pos];
				break;
			}
			pos++;
		}
		vector<Object>::iterator sIter2 = m_callBackInvokes.begin();
		for(; sIter2 != m_callBackInvokes.end(); ++sIter2){
			if(pInvoke == *sIter2){
				m_callBackInvokes.erase(sIter2);
				break;
			}
		}
	}

	//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	FCSocketArray::FCSocketArray(){
	}

	FCSocketArray::~FCSocketArray(){
	}

	void FCSocketArray::addSocket(int socketID){
		vector<int>::iterator sIter = m_sockets.begin();
		for(; sIter != m_sockets.end(); ++sIter){
			if(socketID == *sIter)
			{
				return;
			}
		}
		m_sockets.push_back(socketID);
	}

	void FCSocketArray::getSocketList(vector<int> *socketList){
		vector<int>::iterator sIter = m_sockets.begin();
		for(; sIter != m_sockets.end(); ++sIter){
			socketList->push_back(*sIter);
		}
	}

	void FCSocketArray::removeSocket(int socketID){
		vector<int>::iterator sIter = m_sockets.begin();
		for(; sIter != m_sockets.end(); ++sIter){
			if(socketID == *sIter){
				m_sockets.erase(sIter);
				return;
			}
		}
	}

	//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	int FCServerService::closeServer(int socketID){
		return FCServerSockets::close(socketID);
	}

	void FCServerService::registerCallBack(MessageCallBack callBack){
		FCServerSockets::registerRecv(callBack);
	}

	void FCServerService::registerLog(WriteLogCallBack callBack){
		FCServerSockets::registerLog(callBack);
	}

	int FCServerService::sendByServer(int socketID, const char *str, int len){
		return FCServerSockets::send(socketID, str, len);
	}

	int FCServerService::startServer(int type, int port){
		return FCServerSockets::start(type, port);
	}

	//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

	FCServerService::FCServerService(){
		m_compressType = COMPRESSTYPE_GZIP;
		m_groupID = 0;
		m_serviceID = 0;
		m_sessionID = 0;
	}

	FCServerService::~FCServerService(){
		m_lock.lock();
		map<int, FCMessageListener*>::iterator sIter = m_listeners.begin();
		for(; sIter != m_listeners.end(); ++sIter){
			delete sIter->second;
		}
		m_listeners.clear();
		m_waitMessages.clear();
		m_lock.unLock();
	}

	int FCServerService::getCompressType(){
		return 1;
	}

	void FCServerService::setCompressType(int compressType){
		m_compressType = compressType;
	}

	long long FCServerService::getDownFlow(){
		return m_downFlow;
	}

	void FCServerService::setDownFlow(long long downFlow){
		m_downFlow = downFlow;
	}

	int FCServerService::getGroupID(){
		return m_groupID;
	}

	void FCServerService::setGroupID(int groupID){
		m_groupID = groupID;
	}

	int FCServerService::getServiceID(){
		return m_serviceID;
	}

	void FCServerService::setServiceID(int serviceID){
		m_serviceID = serviceID;
	}

	int FCServerService::getSessionID(){
		return m_sessionID;
	}

	void FCServerService::setSessionID(int sessionID){
		m_sessionID = sessionID;
	}

	long long FCServerService::getUpFlow(){
		return m_upFlow;
	}

	void FCServerService::setUpFlow(long long upFlow){
		m_upFlow = upFlow;
	}

	/////////////////////////////////////////////////////////////////////////////////////////////////////////

	void FCServerService::addService(FCServerService *service){
		if (!m_messageCallBack){
			MessageCallBack messageCallBack = &callBack;
			m_messageCallBack = messageCallBack;
			registerCallBack(messageCallBack);
			WriteLogCallBack writeLogCallBack = &writeServerLog;
			m_writeLogCallBack = writeLogCallBack;
			registerLog(m_writeLogCallBack);
		}
		m_services[service->getServiceID()] = service;
	}

	void FCServerService::callBack(int socketID, int localSID, const char *str, int len){
		m_downFlow += len;
		if (len > 4){
			FCBinary binary;
			binary.write(str, len);
			int head = binary.readInt();
			int groupID = (int)binary.readShort();
			int serviceID = (int)binary.readShort();
			FCServerService *service = 0;
			if(m_services.find(serviceID) != m_services.end()){
				m_services[serviceID]->onCallBack(&binary, socketID, localSID, len);
			}
		}
	}

	int FCServerService::getRequestID(){
		return m_requestID ++;
	}

	void FCServerService::getServices(vector<FCServerService*> *services){
		map<int, FCServerService*>::iterator sIter = m_services.begin();
		for(; sIter != m_services.end(); ++sIter){
			services->push_back(sIter->second);
		}
	}

	int FCServerService::keepAlive(int socketID){
		FCBinary binary;
		binary.initialize(4, false);
		binary.writeInt(4);
		char *str = (char*)binary.getBytes();
		return sendByServer(socketID, str, 4);
	}

	void FCServerService::onCallBack(FCBinary *binary, int socketID, int localSID, int len){
		int headSize = sizeof(int) * 4 + sizeof(short) * 3 + sizeof(byte) * 2;
		int functionID = (int)binary->readShort();
		int sessionID = binary->readInt();
		int requestID = binary->readInt();
		int state = (int)binary->readChar();
		int compressType = (int)binary->readChar();
		int uncBodyLength = binary->readInt();
		int bodyLength = len - headSize;
		char *body = new char[bodyLength];
		memset(body, '\0', bodyLength);
		binary->read(body, bodyLength);
		if (compressType == COMPRESSTYPE_GZIP){
			//uLong cLen = 128;
			//if(uncBodyLength > cLen)
			//{
			//	cLen = uncBodyLength;
			//}
			//char *buffer = new char[cLen];
			//memset(buffer, '\0', cLen);
			//FCStr::GZDeCompress((Byte*)body, bodyLength, (Byte*)buffer, &cLen);
			//delete[] body;
			//body = buffer;
			//uncBodyLength = cLen;
		}
		FCMessage message(getGroupID(), getServiceID(), functionID, sessionID, requestID, socketID, state, compressType, uncBodyLength, body);
		onReceive(&message);
		onWaitMessageHandle(&message);
		if(message.m_body){
			delete[] body;
		}
		body = 0;
	}

	void FCServerService::onClientClose(int socketID, int localSID){
	}

	void FCServerService::onReceive(FCMessage *message){
		m_lock.lock();
		m_compressTypes[message->m_socketID] = message->m_compressType;
		m_lock.unLock();
	}

	void FCServerService::onWaitMessageHandle(FCMessage *message){
		if((int)m_waitMessages.size() > 0){
			m_lock.lock();
			map<int, FCMessage*>::iterator sIter = m_waitMessages.find(message->m_requestID);
			if(sIter != m_waitMessages.end()){
				FCMessage *waitMessage = sIter->second;
				waitMessage->copy(message);
				char *newBody = new char[message->m_bodyLength];
				for (int i = 0; i < message->m_bodyLength; i++){
					newBody[i] = message->m_body[i];
				}
				waitMessage->m_body = newBody;
			}
			m_lock.unLock();
		}
	}

	void FCServerService::registerWait(int requestID, FCMessage *message){
		m_lock.lock();
		m_waitMessages[requestID] = message;
		m_lock.unLock();
	}

	void FCServerService::registerListener(int requestID, ListenerMessageCallBack callBack, Object pInvoke){
		m_lock.lock();
		FCMessageListener *listener = 0;
		map<int, FCMessageListener*>::iterator sIter = m_listeners.find(requestID);
		if(sIter != m_listeners.end()){
			listener = sIter->second;
		}
		else{
			listener = new FCMessageListener;
			m_listeners[requestID] = listener;
		}
		listener->add(callBack, pInvoke);
		m_lock.unLock();
	}

	int FCServerService::send(FCMessage *message){
		FCBinary binary;
		char *buffer = 0;
		const char *body = message->m_body;
		int bodyLength = message->m_bodyLength;
		int uncBodyLength = bodyLength;
		m_lock.lock();
		map<int, int>::iterator sIter = m_compressTypes.find(message->m_socketID);
		if(sIter != m_compressTypes.end()){
			message->m_compressType = sIter->second;
		}
		m_lock.unLock();
		if (message->m_compressType == COMPRESSTYPE_GZIP){
			//uLong cLen = bodyLength;
			//buffer = new char[cLen];
			//memset(buffer, '\0', cLen);
			////FCStr::GZCompress((Byte*)body, bodyLength, (Byte*)buffer, &cLen);
			//body = buffer;
			//bodyLength = (int)cLen;
		}
		int len = sizeof(int) * 4 + bodyLength + sizeof(short) * 3 + sizeof(char) * 2;
		binary.initialize(len, false);
		binary.writeInt(len);
		binary.writeShort((short)message->m_groupID);
		binary.writeShort((short)message->m_serviceID);
		binary.writeShort((short)message->m_functionID);
		binary.writeInt(message->m_sessionID);
		binary.writeInt(message->m_requestID);
		binary.writeChar((char)message->m_state);
		binary.writeChar((char)message->m_compressType);
		binary.writeInt(uncBodyLength);
		binary.write(body, bodyLength);
		char *str = (char*)binary.getBytes();
		if(buffer){
			delete[] buffer;
			buffer = 0;
		}
		int ret = sendByServer(message->m_socketID, str, len);
		m_upFlow += ret;
		return ret;
	}

	void FCServerService::sendToListener(FCMessage *message){
		FCMessageListener *listener = 0;
		Object pInvoke = 0;
		m_lock.lock();
		map<int, FCMessageListener*>::iterator sIter = m_listeners.find(message->m_requestID);
		if(sIter != m_listeners.end()){
			listener = sIter->second;
		}
		m_lock.unLock();
		if(listener){
			listener->callBack(message);
		}
	}

	void FCServerService::unRegisterListener(int requestID){
		m_lock.lock();
		map<int, FCMessageListener*>::iterator sIter = m_listeners.find(requestID);
		if(sIter != m_listeners.end()){
			delete sIter->second;
			m_listeners.erase(sIter);
		}
		m_lock.unLock();
	}

	void FCServerService::unRegisterListener(int requestID, ListenerMessageCallBack callBack){
		m_lock.lock();
		map<int, FCMessageListener*>::iterator sIter = m_listeners.find(requestID);
		if(sIter != m_listeners.end()){
			sIter->second->remove(callBack);
		}
		m_lock.unLock();
	}

	void FCServerService::unRegisterWait(int requestID){
		m_lock.lock();
		map<int, FCMessage*>::iterator sIter = m_waitMessages.find(requestID);
		if(sIter != m_waitMessages.end()){
			m_waitMessages.erase(sIter);
		}
		m_lock.unLock();
	}

	int FCServerService::waitMessage(int requestID, int timeout){
		int state = 0;
		while(timeout > 0){
			m_lock.lock();
			map<int, FCMessage*>::iterator sIter = m_waitMessages.find(requestID);
			if(sIter != m_waitMessages.end()){
				FCMessage *message = sIter->second;
				if(message->m_bodyLength > 0){
					state = 1;
					m_lock.unLock();
					break;
				}
			}
			else{
				m_lock.unLock();
				break;
			}
			m_lock.unLock();
			timeout -= 10;
			Sleep(10);
		}
		unRegisterWait(requestID);
		return state;
	}

	void FCServerService::writeServerLog(int socketID, int localSID, int state, const char *log){
		if(state == 2){
			map<int, FCServerService*>::iterator sIter = m_services.begin();
			for(; sIter != m_services.end(); ++sIter){
				FCServerService *service = sIter->second;
				service->onClientClose(socketID, localSID);
			}
		}
	}
}

File: facecat_java/src/facecat/topin/service/FCServerService.java
/*捂脸猫FaceCat框架 v1.0
 1.创始人-矿洞程序员-上海宁米科技创始人-脉脉KOL-陶德 (微信号:suade1984);
 2.联合创始人-上海宁米科技创始人-袁立涛(微信号:wx627378127);
 3.联合创始人-河北思尔企业管理咨询有限公司合伙人-肖添龙(微信号:xiaotianlong_luu);
 4.联合开发者-陈晓阳(微信号:chenxiaoyangzxy)，助理-朱炜(微信号:cnnic_zhu);
 5.该框架开源协议为BSD，欢迎对我们的创业活动进行各种支持，欢迎更多开发者加入。
 包含C/C++,Java,C#,iOS,MacOS,Linux六个版本的图形和通讯服务框架。
 */
package facecat.topin.service;

import facecat.topin.core.*;
import java.io.*;
import java.util.*;
import facecat.topin.sock.*;

/**
 * 基础服务类
 *
 */
public class FCServerService {

    /**
     * 创建服务
     *
     */
    public FCServerService() {
    }

    /**
     * 析构函数
     *
     */
    protected void finalize() throws Throwable {
        synchronized (m_listeners) {
            m_listeners.clear();
        }
        synchronized (m_waitMessages) {
            m_waitMessages.clear();
        }
    }

    /**
     * 无压缩
     *
     */
    public static int COMPRESSTYPE_NONE = 0;

    /**
     * GZIP压缩
     *
     */
    public static int COMPRESSTYPE_GZIP = 1;

    /**
     * 压缩类型
     */
    protected HashMap<Integer, Integer> m_compressTypes = new HashMap<Integer, Integer>();

    /**
     * 监听者集合
     *
     */
    private java.util.HashMap<Integer, FCMessageListener> m_listeners = new java.util.HashMap<Integer, FCMessageListener>();

    /**
     * 请求ID
     *
     */
    private static int m_requestID = 10000;

    /**
     * 所有的服务
     *
     */
    private static java.util.HashMap<Integer, FCServerService> m_services = new java.util.HashMap<Integer, FCServerService>();

    /**
     * 等待消息队列
     *
     */
    private java.util.HashMap<Integer, FCMessage> m_waitMessages = new java.util.HashMap<Integer, FCMessage>();

    private int m_compressType = COMPRESSTYPE_GZIP;

    /**
     * 获取或设置压缩类型
     *
     */
    public final int getCompressType() {
        return m_compressType;
    }

    public final void setCompressType(int value) {
        m_compressType = value;
    }

    protected static long m_downFlow;

    /**
     * 获取或设置下载流量
     *
     */
    public static long getDownFlow() {
        return FCServerService.m_downFlow;
    }

    public static void setDownFlow(long value) {
        FCServerService.m_downFlow = value;
    }

    private int m_groupID = 0;

    /**
     * 获取或设置组ID
     *
     */
    public final int getGroupID() {
        return m_groupID;
    }

    public final void setGroupID(int value) {
        m_groupID = value;
    }

    private boolean m_isDeleted = false;

    /**
     * 获取对象是否已被销毁
     *
     */
    public final boolean isDeleted() {
        return m_isDeleted;
    }

    private int m_serviceID = 0;

    /**
     * 获取或设置服务的ID
     *
     */
    public final int getServiceID() {
        return m_serviceID;
    }

    public final void setServiceID(int value) {
        m_serviceID = value;
    }

    private int m_sessionID = 0;

    /**
     * 获取或设置登录ID
     *
     */
    public final int getSessionID() {
        return m_sessionID;
    }

    public final void detSessionID(int value) {
        m_sessionID = value;
    }

    private int m_socketID = 0;

    public final int getSocketID() {
        return m_socketID;
    }

    public final void setSocketID(int socketID) {
        m_socketID = socketID;
    }

    protected static long m_upFlow;

    /**
     * 获取或设置上传流量
     *
     */
    public static long getUpFlow() {
        return m_upFlow;
    }

    public static void getUpFlow(long value) {
        m_upFlow = value;
    }

    /**
     * 关闭
     *
     * @param socketID 连接ID
     * @return 状态
     */
    public static int closeServer(int socketID) {
        return FCServerSockets.close(socketID);
    }

    /**
     * 发送消息
     *
     * @param socketID 连接ID
     * @param str 数据
     * @param len 长度
     * @return 状态
     */
    public static int sendByServer(int socketID, int localSID, byte[] str, int len) {
        return FCServerSockets.send(socketID, localSID, str, len);
    }

    /**
     * 启动服务
     *
     * @param port 端口号
     * @return 状态
     */
    public static int startServer(int port) {
        FCServerSockets.setListener(new FCServerSocketListener());
        return FCServerSockets.start(0, port);
    }

    /**
     * 添加服务
     *
     * @param service 服务
     */
    public static void addService(FCServerService service) {
        m_services.put(service.getServiceID(), service);
    }

    /**
     * 回调函数
     *
     * @param socketID 连接ID
     * @param localSID 本地连接ID
     * @param str 数据
     * @param len 长度
     */
    public static void callBack(int socketID, int localSID, byte[] str, int len) {
        m_downFlow += len;
        try {
            if (len > 4) {
                FCBinary br = new FCBinary();
                br.write(str, len);
                int head = br.readInt();
                int groupID = br.readShort();
                int serviceID = br.readShort();
                FCServerService service = null;
                if (m_services.containsKey(serviceID)) {
                    m_services.get(serviceID).onCallBack(br, socketID, localSID, len);
                }
                br.close();
            }
        } catch (Exception ex) {
            System.out.println(ex.getMessage() + "\r\n" + ex.getStackTrace());
        }
    }

    /**
     * 销毁对象
     *
     */
    public void delete() {
        if (!m_isDeleted) {
            m_listeners.clear();
            m_isDeleted = true;
        }
    }

    /**
     * 获取请求ID
     *
     * @return 请求ID
     */
    public static int getRequestID() {
        return m_requestID++;
    }

    /**
     * 获取所有的服务
     *
     * @param services 服务列表
     */
    public static void getServices(ArrayList<FCServerService> services) {
        for (FCServerService service : m_services.values()) {
            services.add(service);
        }
    }

    /**
     * 保持活跃
     *
     * @param socketID 套接字ID
     * @return 状态
     */
    public int keepAlive(int socketID) {
        int ret = -1;
        try {
            FCBinary bw = new FCBinary();
            bw.writeInt((int) 4);
            ret = sendByServer(socketID, m_socketID, bw.getBytes(), 4);
            bw.close();
        } catch (Exception ex) {

        }
        return ret;
    }

    /**
     * 收到消息
     *
     * @param br 流
     * @param socketID 套接字ID
     * @param localSID 本地套接字ID
     * @param len 长度
     */
    public void onCallBack(FCBinary br, int socketID, int localSID, int len) {
        try {
            int headSize = 4 * 4 + 2 * 3 + 1 * 2;
            int functionID = br.readShort();
            int sessionID = br.readInt();
            int requestID = br.readInt();
            int state = (int) br.readChar();
            int compressType = (int) br.readChar();
            int bodyLength = br.readInt();
            byte[] body = new byte[len - headSize];
            br.readBytes(body);
            FCMessage message = null;
            byte[] unzipBody = null;
            if (compressType == COMPRESSTYPE_GZIP) {
                //unzipBody = CStrA.UnGZip(body);
                bodyLength = unzipBody.length;
                message = new FCMessage(getGroupID(), getServiceID(), functionID, sessionID, requestID, socketID, state,
                        compressType, bodyLength, unzipBody);
                // TODO...
            } else {
                message = new FCMessage(getGroupID(), getServiceID(), functionID, sessionID, requestID, socketID, state,
                        compressType, bodyLength, body);
            }

            onReceive(message);
            onWaitMessageHandle(message);
            body = null;
            unzipBody = null;
        } catch (Exception ex) {

        }
    }

    /**
     * 客户端退出方法
     *
     */
    public void onClientClose(int socketID, int localSID) {
        if (m_compressTypes.containsKey(socketID)) {
            m_compressTypes.remove(socketID);
        }
    }

    /**
     * 接收数据
     *
     * @param message 消息
     */
    public void onReceive(FCMessage message) {
        synchronized (m_compressTypes) {
            m_compressTypes.put(message.m_socketID, message.m_compressType);
        }
    }

    /**
     * 等待消息的处理
     *
     * @param message 消息
     */
    public void onWaitMessageHandle(FCMessage message) {
        if (m_waitMessages.size() > 0) {
            synchronized (m_waitMessages) {
                if (m_waitMessages.containsKey(message.m_requestID)) {
                    FCMessage waitMessage = m_waitMessages.get(message.m_requestID);
                    waitMessage.copy(message);
                }
            }
        }
    }

    /**
     * 注册数据监听
     *
     * @param requestID 请求ID
     * @param callBack 回调函数
     */
    public void registerListener(int requestID, FCListenerMessageCallBack callBack) {
        synchronized (m_listeners) {
            FCMessageListener listener = null;
            if (!m_listeners.containsKey(requestID)) {
                listener = new FCMessageListener();
                m_listeners.put(requestID, listener);
            } else {
                listener = m_listeners.get(requestID);
            }
            listener.add(callBack);
        }
    }

    /**
     * 注册等待
     *
     * @param requestID 请求ID
     * @param message 消息
     */
    public void registerWait(int requestID, FCMessage message) {
        synchronized (m_waitMessages) {
            m_waitMessages.put(requestID, message);
        }
    }

    /**
     * 发送消息
     *
     * @param message 消息
     */
    public int send(FCMessage message) {
        int ret = -1;
        try {
            FCBinary bw = new FCBinary();
            byte[] body = message.m_body;
            int bodyLength = message.m_bodyLength;
            int uncBodyLength = bodyLength;
            synchronized (m_compressTypes) {
                if (m_compressTypes.containsKey(message.m_socketID)) {
                    message.m_compressType = m_compressTypes.get(message.m_socketID);
                }
            }
            byte[] zipBody = null;
            if (message.m_compressType == COMPRESSTYPE_GZIP) {
                zipBody = FCStr.gZip(body);
                bodyLength = zipBody.length;
            }
            int len = 4 * 4 + bodyLength + 2 * 3 + 1 * 2;
            bw.writeInt(len);
            bw.writeShort((short) message.m_groupID);
            bw.writeShort((short) message.m_serviceID);
            bw.writeShort((short) message.m_functionID);
            bw.writeInt(message.m_sessionID);
            bw.writeInt(message.m_requestID);
            bw.writeChar((char) message.m_state);
            bw.writeChar((char) message.m_compressType);
            bw.writeInt(uncBodyLength);
            if (message.m_compressType == COMPRESSTYPE_GZIP) {
                bw.writeBytes(zipBody);
            } else {
                bw.writeBytes(body);
            }
            byte[] bytes = bw.getBytes();
            ret = sendByServer(message.m_socketID, m_socketID, bytes, bytes.length);
            bw.close();
        } catch (Exception ex) {

        }
        return ret;
    }

    /**
     * 发送到监听者
     *
     * @param message 消息
     */
    public void sendToListener(FCMessage message) {
        FCMessageListener listener = null;
        synchronized (m_listeners) {
            if (m_listeners.containsKey(message.m_requestID)) {
                listener = m_listeners.get(message.m_requestID);
            }
        }
        if (listener != null) {
            listener.callBack(message);
        }
    }

    /**
     * 取消注册数据监听
     *
     * @param requestID 请求ID
     */
    public void unRegisterListener(int requestID) {
        synchronized (m_listeners) {
            m_listeners.remove(requestID);
        }
    }

    /**
     * 取消注册监听
     *
     * @param requestID 请求ID
     * @param callBack 回调函数
     */
    public void unRegisterListener(int requestID, FCListenerMessageCallBack callBack) {
        synchronized (m_listeners) {
            if (m_listeners.containsKey(requestID)) {
                m_listeners.get(requestID).remove(callBack);
            }
        }
    }

    /**
     * 取消注册等待
     *
     * @param requestID 请求ID
     */
    public void unRegisterWait(int requestID) {
        synchronized (m_waitMessages) {
            if (m_waitMessages.containsKey(requestID)) {
                m_waitMessages.remove(requestID);
            }
        }
    }

    /**
     * 等待消息
     *
     * @param requestID 请求ID
     * @param timeout 超时
     * @return 状态
     */
    public int waitMessage(int requestID, int timeout) {
        int state = 0;
        try {
            while (timeout > 0) {
                synchronized (m_waitMessages) {
                    if (m_waitMessages.containsKey(requestID)) {
                        if (m_waitMessages.get(requestID).m_bodyLength > 0) {
                            state = 1;
                            break;
                        }
                    } else {
                        break;
                    }
                }
                timeout -= 10;
                Thread.sleep(10);
            }
            unRegisterWait(requestID);
        } catch (Exception ex) {
        }
        return state;
    }

    /**
     * 等待消息
     *
     * @param socketID 目标SocketID
     * @param localSID 本地SocketID
     * @param state 状态
     * @param log 日志
     * @return 状态
     */
    public static void writeLog(int socketID, int localSID, int state, String log) {
        if (state == 2) {
            for (FCServerService service : m_services.values()) {
                service.onClientClose(socketID, localSID);
            }
        }
    }
}
